File: ffi/bindings/java/examples/src/main/java/io/stepfunc/rodbus/examples/ServerExample.java
package io.stepfunc.rodbus.examples;

import static org.joou.Unsigned.*;

import io.stepfunc.rodbus.*;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.time.Duration;
import java.util.Arrays;
import java.util.List;

import io.stepfunc.rodbus.Runtime;
import org.joou.UByte;
import org.joou.UShort;

class TestLogger implements Logger {

    @Override
    public void onMessage(LogLevel level, String message) {
        System.out.print(message);
    }
}

// ANCHOR: write_handler
class ExampleWriteHandler implements WriteHandler {
    @Override
    public WriteResult writeSingleCoil(UShort index, boolean value, Database database) {
        if(database.updateCoil(index, value)) {
            return WriteResult.successInit();
        } else {
            return WriteResult.exceptionInit(ModbusException.ILLEGAL_DATA_ADDRESS);
        }
    }

    @Override
    public WriteResult writeSingleRegister(UShort index, UShort value, Database database) {
        if(database.updateHoldingRegister(index, value)) {
            return WriteResult.successInit();
        } else {
            return WriteResult.exceptionInit(ModbusException.ILLEGAL_DATA_ADDRESS);
        }
    }

    @Override
    public WriteResult writeMultipleCoils(UShort start, List<BitValue> it, Database database) {
        WriteResult result = WriteResult.successInit();

        for(BitValue bit : it) {
            if(!database.updateCoil(bit.index, bit.value)) {
                result = WriteResult.exceptionInit(ModbusException.ILLEGAL_DATA_ADDRESS);
            }
        }

        return result;
    }

    @Override
    public WriteResult writeMultipleRegisters(UShort start, List<RegisterValue> it, Database database) {
        WriteResult result = WriteResult.successInit();

        for(RegisterValue reg : it) {
            if(!database.updateHoldingRegister(reg.index, reg.value)) {
                result = WriteResult.exceptionInit(ModbusException.ILLEGAL_DATA_ADDRESS);
            }
        }

        return result;
    }
}
// ANCHOR_END: write_handler

// ANCHOR: auth_handler
class TestAuthorizationHandler implements AuthorizationHandler
{
    public Authorization readCoils(UByte unitId, AddressRange range, String role) {
        return Authorization.ALLOW;
    }

    public Authorization readDiscreteInputs(UByte unitId, AddressRange range, String role) {
        return Authorization.ALLOW;
    }

    public Authorization readHoldingRegisters(UByte unitId, AddressRange range, String role) {
        return Authorization.ALLOW;
    }

    public Authorization readInputRegisters(UByte unitId, AddressRange range, String role) {
        return Authorization.ALLOW;
    }

    public Authorization writeSingleCoil(UByte unitId, UShort idx, String role) {
        return Authorization.ALLOW;
    }

    public Authorization writeSingleRegister(UByte unitId, UShort idx, String role) {
        return Authorization.DENY;
    }

    public Authorization writeMultipleCoils(UByte unitId, AddressRange range, String role) {
        return Authorization.DENY;
    }

    public Authorization writeMultipleRegisters(UByte unitId, AddressRange range, String role) {
        return Authorization.DENY;
    }
}
// ANCHOR_END: auth_handler

public class ServerExample {
    public static void main(String[] args) throws Exception {
        // initialize logging with the default configuration
        Logging.configure(new LoggingConfig(), new ConsoleLogger());

        // initialize the runtime
        RuntimeConfig runtimeConfig = new RuntimeConfig().withNumCoreThreads(ushort(4));
        Runtime runtime = new Runtime(runtimeConfig);

        // create the device map
        // ANCHOR: device_map_init
        DeviceMap map = new DeviceMap();
        map.addEndpoint(ubyte(1), new ExampleWriteHandler(), db -> {
            for(int i = 0; i < 10; i++) {
                db.addCoil(ushort(i), false);
                db.addDiscreteInput(ushort(i), false);
                db.addHoldingRegister(ushort(i), ushort(0));
                db.addInputRegister(ushort(i), ushort(0));
            }
        });
        // ANCHOR_END: device_map_init

        if (args.length != 1)
        {
            System.out.println("you must specify a transport type");
            System.out.println("usage: server_example <channel> (tcp, rtu, tls-ca, tls-self-signed)");
            System.exit(-1);
        }

        // initialize a Modbus client channel
        Server server = createServer(args[0], runtime, map);

        try {
            run(server);
        } finally {
            runtime.shutdown();
        }
    }

    private static Server createServer(String type, Runtime runtime, DeviceMap map) {
        switch (type)
        {
            case "tcp": return createTcpServer(runtime, map);
            case "rtu": return createRtuServer(runtime, map);
            case "tls-ca": return createTlsServer(runtime, map, getCaTlsConfig());
            case "tls-self-signed": return createTlsServer(runtime, map, getSelfSignedTlsConfig());
            default:
                System.out.println("unknown server type: " + type);
                System.exit(-1);
                return null;
        }
    }

    private static Server createTcpServer(Runtime runtime, DeviceMap map) {
        // ANCHOR: tcp_server_create
        Server server = Server.createTcp(runtime, "127.0.0.1", ushort(502), ushort(100), map, DecodeLevel.nothing());
        // ANCHOR_END: tcp_server_create

        return server;
    }

    private static Server createRtuServer(Runtime runtime, DeviceMap map) {
        // ANCHOR: rtu_server_create
        Server server = Server.createRtu(runtime, "/dev/ttySIM1", new SerialPortSettings(), map, DecodeLevel.nothing());
        // ANCHOR_END: rtu_server_create

        return server;
    }

    private static Server createTlsServer(Runtime runtime, DeviceMap map, TlsServerConfig tlsConfig) {
        // ANCHOR: tls_server_create
        Server server = Server.createTlsWithAuthz(runtime, "127.0.0.1", ushort(802), ushort(10), map, tlsConfig, new TestAuthorizationHandler(), DecodeLevel.nothing());
        // ANCHOR_END: tls_server_create

        return server;
    }

    private static TlsServerConfig getCaTlsConfig() {
        // ANCHOR: tls_ca_chain_config
        TlsServerConfig tlsConfig = new TlsServerConfig(
                "./certs/ca_chain/ca_cert.pem",
                "./certs/ca_chain/server_cert.pem",
                "./certs/ca_chain/server_key.pem",
                "" // no password
        );
        // ANCHOR_END: tls_ca_chain_config

        return tlsConfig;
    }

    private static TlsServerConfig getSelfSignedTlsConfig() {
        // ANCHOR: tls_self_signed_config
        TlsServerConfig tlsConfig = new TlsServerConfig(
                "./certs/self_signed/entity1_cert.pem",
                "./certs/self_signed/entity2_cert.pem",
                "./certs/self_signed/entity2_key.pem",
                "" // no password
        ).withCertificateMode(CertificateMode.SELF_SIGNED);
        // ANCHOR_END: tls_self_signed_config

        return tlsConfig;
    }

    public static void run(Server server) throws Exception {
        boolean coilValue = false;
        boolean discreteInputValue = false;
        int holdingRegisterValue = 0;
        int inputRegisterValue = 0;

        // Handle user input
        final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        while (true) {
            final String line = reader.readLine();
            switch (line) {
                case "x":
                    return;
                case "uc":
                {
                    // ANCHOR: update_coil
                    coilValue = !coilValue;
                    final boolean pointValue = coilValue;
                    server.updateDatabase(ubyte(1), db -> {
                        for(int i = 0; i < 10; i++) {
                            db.updateCoil(ushort(i), pointValue);
                        }
                    });
                    // ANCHOR_END: update_coil
                    break;
                }
                case "udi":
                {
                    discreteInputValue = !discreteInputValue;
                    final boolean pointValue = discreteInputValue;
                    server.updateDatabase(ubyte(1), db -> {
                        for(int i = 0; i < 10; i++) {
                            db.updateDiscreteInput(ushort(i), pointValue);
                        }
                    });
                    break;
                }
                case "uhr":
                {
                    holdingRegisterValue += 1;
                    final UShort pointValue = ushort(holdingRegisterValue);
                    server.updateDatabase(ubyte(1), db -> {
                        for(int i = 0; i < 10; i++) {
                            db.updateHoldingRegister(ushort(i), pointValue);
                        }
                    });
                    break;
                }
                case "uir":
                {
                    inputRegisterValue += 1;
                    final UShort pointValue = ushort(inputRegisterValue);
                    server.updateDatabase(ubyte(1), db -> {
                        for(int i = 0; i < 10; i++) {
                            db.updateInputRegister(ushort(i), pointValue);
                        }
                    });
                    break;
                }
                default:
                    System.out.println("Unknown command");
                    break;
            }
        }
    }
}

File: ffi/bindings/c/server_example.cpp
#include "rodbus.hpp"

#include <chrono>
#include <cstring>
#include <iostream>
#include <string>

class Logger : public rodbus::Logger {
    void on_message(rodbus::LogLevel level, const char* message) override
    {
        std::cout << message;
    }
};

// ANCHOR: write_handler
class WriterHandler : public rodbus::WriteHandler
{
    rodbus::WriteResult write_single_coil(uint16_t index, bool value, rodbus::Database& db) override
    {
        if(db.update_coil(index, value))
        {
            return rodbus::WriteResult::success_init();
        }
        else
        {
            return rodbus::WriteResult::exception_init(rodbus::ModbusException::illegal_data_address);
        }
    }

    rodbus::WriteResult write_single_register(uint16_t index, uint16_t value, rodbus::Database& db) override
    {
        if(db.update_holding_register(index, value))
        {
            return rodbus::WriteResult::success_init();
        }
        else
        {
            return rodbus::WriteResult::exception_init(rodbus::ModbusException::illegal_data_address);
        }
    }

    rodbus::WriteResult write_multiple_coils(uint16_t index, rodbus::BitValueIterator& it, rodbus::Database& db) override
    {
        auto result = rodbus::WriteResult::success_init();
        while (it.next())
        {
            const auto bit = it.get();
            if (!db.update_coil(bit.index, bit.value))
            {
                result = rodbus::WriteResult::exception_init(rodbus::ModbusException::illegal_data_address);
            }
        }
        return result;
    }

    rodbus::WriteResult write_multiple_registers(uint16_t index, rodbus::RegisterValueIterator& it, rodbus::Database& db) override
    {
        auto result = rodbus::WriteResult::success_init();
        while (it.next())
        {
            const auto register_value = it.get();
            if (!db.update_holding_register(register_value.index, register_value.value))
            {
                result = rodbus::WriteResult::exception_init(rodbus::ModbusException::illegal_data_address);
            }
        }
        return result;
    }
};
// ANCHOR_END: write_handler

// ANCHOR: auth_handler
class AuthorizationHandler : public rodbus::AuthorizationHandler
{
    rodbus::Authorization read_coils(uint8_t unit_id, const rodbus::AddressRange& range, const char* role) override
    {
        return rodbus::Authorization::allow;
    }

    rodbus::Authorization read_discrete_inputs(uint8_t unit_id, const rodbus::AddressRange &range, const char *role) override
    {
        return rodbus::Authorization::allow;
    }

    rodbus::Authorization read_holding_registers(uint8_t unit_id, const rodbus::AddressRange &range, const char *role) override
    {
        return rodbus::Authorization::allow;
    }

    rodbus::Authorization read_input_registers(uint8_t unit_id, const rodbus::AddressRange &range, const char *role) override
    {
        return rodbus::Authorization::allow;
    }

    rodbus::Authorization write_single_coil(uint8_t unit_id, uint16_t idx, const char *role) override
    {
        return rodbus::Authorization::deny;
    }

    rodbus::Authorization write_single_register(uint8_t unit_id, uint16_t idx, const char *role) override
    {
        return rodbus::Authorization::deny;
    }

    rodbus::Authorization write_multiple_coils(uint8_t unit_id, const rodbus::AddressRange &range, const char *role) override
    {
        return rodbus::Authorization::deny;
    }

    rodbus::Authorization write_multiple_registers(uint8_t unit_id, const rodbus::AddressRange &range, const char *role) override
    {
        return rodbus::Authorization::deny;
    }
};
// ANCHOR_END: auth_handler

int run_server(rodbus::Server& server)
{
    // state passed to the update callbacks
    auto coil_value = false;
    auto discrete_input_value = false;
    auto holding_register_value = 0;
    auto input_register_value = 0;

    while (true) {
        std::string cmd;
        std::getline(std::cin, cmd);

        if (cmd == "x") {
            return 0;
        }
        else if (cmd == "ed") {
            // enable decoding
            server.set_decode_level(
                rodbus::DecodeLevel(rodbus::AppDecodeLevel::data_values, rodbus::FrameDecodeLevel::header, rodbus::PhysDecodeLevel::length));
        }
        else if (cmd == "dd") {
            // disable decoding
            server.set_decode_level(rodbus::DecodeLevel::nothing());
        }
        else if (cmd == "uc") {
            // ANCHOR: update_coil
            auto transaction = rodbus::functional::database_callback([&](rodbus::Database& db) {
                coil_value = !coil_value;

                for (uint16_t i = 0; i < 10; ++i) {
                    db.update_coil(i, coil_value);
                }
            });
            server.update_database(1, transaction);
            // ANCHOR_END: update_coil
        }
        else if (cmd == "udi") {
            auto transaction = rodbus::functional::database_callback([&](rodbus::Database& db) {
                discrete_input_value = !discrete_input_value;

                for (uint16_t i = 0; i < 10; ++i) {
                    db.update_discrete_input(i, discrete_input_value);
                }
            });
            server.update_database(1, transaction);
        }
        else if (cmd == "uhr") {
            auto transaction = rodbus::functional::database_callback([&](rodbus::Database& db) {
                ++holding_register_value;

                for (uint16_t i = 0; i < 10; ++i) {
                    db.update_holding_register(i, holding_register_value);
                }
            });
            server.update_database(1, transaction);
        }
        else if (cmd == "uir") {
            auto transaction = rodbus::functional::database_callback([&](rodbus::Database& db) {
                ++input_register_value;

                for (uint16_t i = 0; i < 10; ++i) {
                    db.update_input_register(i, input_register_value);
                }
            });
            server.update_database(1, transaction);
        }
        else {
            std::cout << "unknown command: " << cmd << std::endl;
        }
    }
}

rodbus::DeviceMap create_device_map()
{
    // create the device map
    // ANCHOR: device_map_init
    auto device_map = rodbus::DeviceMap();
    auto init_transaction = rodbus::functional::database_callback([](rodbus::Database& db) {
        for (uint16_t i = 0; i < 10; ++i)
        {
            db.add_coil(i, false);
            db.add_discrete_input(i, false);
            db.add_holding_register(i, 0);
            db.add_input_register(i, 0);
        }
    });
    device_map.add_endpoint(1, std::make_unique<WriterHandler>(), init_transaction);
    // ANCHOR_END: device_map_init

    return device_map;
}

int run_tcp_server(rodbus::Runtime& runtime)
{
    auto device_map = create_device_map();

    // ANCHOR: tcp_server_create
    auto server = rodbus::Server::create_tcp(runtime, "127.0.0.1", 502, rodbus::AddressFilter::any(), 100, device_map, rodbus::DecodeLevel::nothing());
    // ANCHOR_END: tcp_server_create

    return run_server(server);
}

int run_rtu_server(rodbus::Runtime& runtime)
{
    auto device_map = create_device_map();

    // ANCHOR: rtu_server_create
    auto server = rodbus::Server::create_rtu(runtime, "/dev/ttySIM1", rodbus::SerialPortSettings(), device_map, rodbus::DecodeLevel::nothing());
    // ANCHOR_END: rtu_server_create

    return run_server(server);
}

int run_tls_server(rodbus::Runtime& runtime, const rodbus::TlsServerConfig& tls_config)
{
    auto device_map = create_device_map();

    // ANCHOR: tls_server_create
    auto server = rodbus::Server::create_tls_with_authz(runtime, "127.0.0.1", 802, rodbus::AddressFilter::any(), 100, device_map, tls_config, std::make_unique<AuthorizationHandler>(), rodbus::DecodeLevel::nothing());
    // ANCHOR_END: tls_server_create

    return run_server(server);
}

rodbus::TlsServerConfig get_tls_ca_config()
{
    // ANCHOR: tls_ca_chain_config
    auto tls_config = rodbus::TlsServerConfig(
        "./certs/ca_chain/ca_cert.pem",
        "./certs/ca_chain/server_cert.pem",
        "./certs/ca_chain/server_key.pem",
        "" // no password
    );
    // ANCHOR_END: tls_ca_chain_config

    return tls_config;
}

rodbus::TlsServerConfig get_tls_self_signed_config()
{
    // ANCHOR: tls_self_signed_config
    auto tls_config = rodbus::TlsServerConfig(
        "./certs/self_signed/entity1_cert.pem",
        "./certs/self_signed/entity2_cert.pem",
        "./certs/self_signed/entity2_key.pem",
        "" // no password
    );
    tls_config.certificate_mode = rodbus::CertificateMode::self_signed;
    // ANCHOR_END: tls_self_signed_config

    return tls_config;
}

int main(int argc, char* argv[])
{
    // initialize logging with the default configuration
    rodbus::Logging::configure(rodbus::LoggingConfig(), std::make_unique<Logger>());

    // initialize the runtime
    auto runtime_config = rodbus::RuntimeConfig();
    runtime_config.num_core_threads = 4;
    auto runtime = rodbus::Runtime(runtime_config);

    if (argc != 2) {
        std::cout << "you must specify a transport type" << std::endl;
        std::cout << "usage: cpp_server_example <channel> (tcp, rtu, tls-ca, tls-self-signed)" << std::endl;
        return -1;
    }

    const auto type = argv[1];

    if (strcmp(type, "tcp") == 0) {
        return run_tcp_server(runtime);
    }
    else if (strcmp(type, "rtu") == 0) {
        return run_rtu_server(runtime);
    }
    else if (strcmp(type, "tls-ca") == 0) {
        return run_tls_server(runtime, get_tls_ca_config());
    }
    else if (strcmp(type, "tls-self-signed") == 0) {
        return run_tls_server(runtime, get_tls_self_signed_config());
    }
    else {
        std::cout << "unknown channel type: " << type << std::endl;
        return -1;
    }
}
